query/dev/system/src/test/groovy/org/codice/ddf/admin/query/dev/system/BundleUtilsTest.groovy

```diff
@@ -0,0 +1,373 @@
+import org.osgi.service.blueprint.reflect.ComponentMetadata
+import org.osgi.service.blueprint.reflect.ReferenceListMetadata
+import org.osgi.service.blueprint.reflect.ReferenceMetadata
+import org.osgi.service.blueprint.reflect.ServiceMetadata
+import spock.lang.Specification
+
+import static org.codice.ddf.admin.query.dev.system.SampleData.*
+
+
+class BundleUtilsTest extends Specification {
+
+    BundleUtils bundleUtils
+    Map<Bundle, Map<String, Bundle>> bundleToPkgDepMap = new HashMap<>()
+    List<Bundle> allMockBundles = []
+    Map<Bundle, MockBlueprintContainer> allBlueprintContainers = new HashMap<>()
+    Map<String, ServiceReference> allServiceRefs = new HashMap<>()
+    Map<ServiceReference, Object> allServices = new HashMap<>()
+    Map<Bundle, List<ServiceReference>> bundleToRegisteredServices = new HashMap<>()
+
+    static final String NO_EXISTING_SERVICE_ID = "no_existing_service_id"
+
+
+    static final
+    def MATCH_PKG = { PackageField pkg, String name, int id -> return pkg.pkgName() == name && pkg.bundleId() == id }
+
+    def setup() {
+        def mockBundleService = Mock(BundleService)
+        mockBundleService.getWiredBundles(_) >> {
+            Bundle bundle -> bundleToPkgDepMap.get(bundle)
+        }
+
+        bundleUtils = new BundleUtils(mockBundleService) {
+            @Override
+            protected <S> S getService(ServiceReference<S> ref) {
+                return allServices.get(ref)
+            }
+
+            @Override
+            protected Optional<BlueprintContainer> getBlueprintContainer(Bundle bundle) {
+                return allBlueprintContainers.containsKey(bundle) ? Optional.of(allBlueprintContainers.get(bundle)) : Optional.empty()
+            }
+
+            @Override
+            protected List<ServiceReference> getServiceReferences(String refInterface, String filter) throws InvalidSyntaxException {
+                //Used filter to pass the ids needed to match
+                List<ServiceReference> matchedRefs = []
+                List<String> refIds = filter.substring(1, filter.length() - 1).split(",")
+                for (String mockRefId : refIds) {
+                    if (allServiceRefs.containsKey(mockRefId)) {
+                        matchedRefs.add(allServiceRefs.get(mockRefId))
+                    }
+                }
+                return matchedRefs
+            }
+
+            @Override
+            protected List<ServiceReference<?>> getRegisteredServices(Bundle bundle) {
+                return bundleToRegisteredServices.containsKey(bundle) ? bundleToRegisteredServices.get(bundle) : Collections.emptyList()
+            }
+
+            @Override
+            protected List<Bundle> getAllBundles() {
+                return allMockBundles
+            }
+        }
+    }
+
+    def "Associates packages between bundles correctly"() {
+        setup:
+        Bundle bundleA = addMockBundle(BUNDLE_A_NAME, BUNDLE_A_ID, BUNDLE_A_LOCATION, [BUNDLE_A_PKG], [:])
```

Comment:
✏️ It looks like each test you are setting a state, in this case what bundles, services, etc are available in a test to the BundleUtils, but can they leak over to other tests? It may help prevent mistakes in the future if we put this information into a new class we can reset each test. Maybe break out the BundleUtils mock and hold all that information in there with setters/getters.

Reply:
I gave this a shot. I tried 3 different ways to refactor the BundleUtils mock into a separate class. I wasn't able to find an elegant way to do this because of the class not having to access to the spock `Mock` class and the number of methods invoking other methods to mock. I don't think this will be valuable from a reusability standpoint anyways. The others tests will more than likely just overriding `getAllBundleFields` and `getBundles`.  I moved the rest of the state into the setup method such that the state is reset before every test to avoid test leakage as you mentioned.